File: examples/go/order-specification/order/repository.go
package order

type InMemoryRepository struct {
	orders []Order
}

func NewInMemoryRepository(orders []Order) InMemoryRepository {
	return InMemoryRepository{
		orders: append([]Order(nil), orders...),
	}
}

func (r InMemoryRepository) FindMatching(specification Specification) []Order {
	matches := make([]Order, 0)

	for _, candidate := range r.orders {
		if specification.IsSatisfiedBy(candidate) {
			matches = append(matches, candidate)
		}
	}

	return matches
}

File: examples/typescript/order-specification/order/repository.ts
import { Order } from "./order";
import { Specification } from "./specification";

export class InMemoryOrderRepository {
  constructor(private readonly orders: Order[]) {}

  findMatching(specification: Specification<Order>): Order[] {
    return this.orders.filter((candidate) => specification.isSatisfiedBy(candidate));
  }
}
